# Assignment 35 - Mar 11' 23 - Statistics Advance 4

### Q1. What is the difference between a t-test and a z-test? Provide an example scenario where you would use each type of test.

| Parameters | T-Test | Z-Test |
| --- | --- | --- |
| Assumption about population standard deviation | Unknown, estimated from sample data | Known |
| Sample size | Small (typically < 30) | Large (typically > 30) |
| Use when comparing | Means of two groups | Means of two groups |
| Example scenario | Comparing the average height of men and women | Testing the effectiveness of a new medical treatment |
| Formula | $t = \frac{\bar{x}_1 - \bar{x}_2}{s_{pooled}\sqrt{\frac{1}{n_1}+\frac{1}{n_2}}}$ | $z = \frac{\bar{x}_1 - \bar{x}_2}{\sqrt{\frac{\sigma_1^2}{n_1}+\frac{\sigma_2^2}{n_2}}}$ |
| Degrees of freedom | $df = n_1 + n_2 - 2$ | Not applicable |
| Critical values | Look up in t-distribution table with degrees of freedom | Look up in standard normal distribution table |
| Hypothesis testing | $H_0$: $\mu_1 = \mu_2$; $H_1$: $\mu_1 \neq \mu_2$ | $H_0$: $\mu_1 = \mu_2$; $H_1$: $\mu_1 \neq \mu_2$ |
| Confidence interval | $\bar{x}_1 - \bar{x}_2 \pm t_{\alpha/2, df} \times s_{pooled}\sqrt{\frac{1}{n_1}+\frac{1}{n_2}}$ | $\bar{x}_1 - \bar{x}_2 \pm z_{\alpha/2}\times \sqrt{\frac{\sigma_1^2}{n_1}+\frac{\sigma_2^2}{n_2}}$ |

* For T-test s_pooled is given as below :
$$ s_{pooled} = \sqrt{\frac{(n_1-1)s_1^2 + (n_2-1)s_2^2}{n_1+n_2-2}} $$

* For T-test s_pooled is given as below :
$$ s_{pooled} = \sqrt{\frac{(n_1-1)s_1^2 + (n_2-1)s_2^2}{n_1+n_2-2}} $$

![t-test vs z-test](https://sixsigmastudyguide.com/wp-content/uploads/2021/08/tz10.png)

#### Example for T- Test :
For example, suppose you want to compare the average test scores of two groups of students who took different types of classes. A t-test would be used to determine whether there is a significant difference between the two groups.
    

#### Example for Z-Test :
suppose you want to compare the average heights of two groups of people from different countries. A z-test would be used to determine whether there is a significant difference between the two groups. Here Sample size >30 and population variance MUST BE KNOWN.

### Q2: Differentiate between one-tailed and two-tailed tests.

|         | One-tailed Test | Two-tailed Test |
|---------|----------------|----------------|
| Definition | A statistical test in which the alternative hypothesis specifies the direction of the effect (either positive or negative) | A statistical test in which the alternative hypothesis does not specify the direction of the effect |
| Directionality | The alternative hypothesis is formulated in a specific direction | The alternative hypothesis is formulated in both directions |
| Hypotheses | Null hypothesis and one-tailed alternative hypothesis | Null hypothesis and two-tailed alternative hypothesis |
| Appropriate Use | When there is prior knowledge or a strong theoretical justification for the direction of the effect | When there is no prior knowledge or theoretical justification for the direction of the effect |
| Example | Testing whether a new drug increases blood pressure | Testing whether a new drug has any effect on blood pressure, regardless of whether it increases or decreases it |

![One vs Two tailed](https://cdn.wallstreetmojo.com/wp-content/uploads/2022/01/One-Tailed-Test-Explained.jpg)

### Q3: Explain the concept of Type 1 and Type 2 errors in hypothesis testing. Provide an example scenario for each type of error.

In hypothesis testing, there are two types of errors that can occur: 
> 1. Type I error
> 2. Type II error.

#### Type I error 
Type I error occurs when the null hypothesis is rejected even though it is actually true. This means that the researcher concludes that there is a significant effect or difference between two groups when in reality there is not. The probability of making a Type I error is denoted as alpha (α) and is set by the researcher prior to conducting the test. A smaller alpha level means a lower probability of making a Type I error.

#### Type II error
Type II error, on the other hand, occurs when the null hypothesis is not rejected even though it is actually false. This means that the researcher concludes that there is no significant effect or difference between two groups when in reality there is. The probability of making a Type II error is denoted as beta (β). The probability of making a Type II error depends on several factors such as the sample size, effect size, and the alpha level.

#### Example

* **Type I error**: Suppose a company produces a batch of 10,000 pens and claims that the average weight of each pen is 5 grams. A researcher randomly selects a sample of 50 pens and finds that the average weight is 5.2 grams with a standard deviation of 0.5 grams. The researcher conducts a hypothesis test with a significance level of 0.05 and concludes that the average weight of the pens is significantly greater than 5 grams. However, in reality, the average weight of the pens is actually 5 grams, and the researcher has made a Type I error by rejecting the null hypothesis.

* **Type II error**: Consider a clinical trial testing the effectiveness of a new drug for treating a certain disease. The null hypothesis is that the drug has no effect on the disease, while the alternative hypothesis is that the drug does have an effect. If the clinical trial fails to reject the null hypothesis, it could be due to the drug having no effect, or it could be due to a Type II error, where the drug actually does have an effect but it was not detected in the study. In this scenario, a Type II error would occur if the new drug is effective, but the clinical trial fails to detect its effectiveness due to a small sample size or other limitations of the study.

![Type 1 and 2](https://datasciencedojo.com/wp-content/uploads/Hypothesis-Testing-Outcomes.png)

### Q4:  Explain Bayes's theorem with an example.

Bayes's theorem is a mathematical formula that provides a way to update the probability of an event occurring based on new information. It is named after Thomas Bayes, an 18th-century British statistician.

Bayes's theorem can be expressed as follows:

$$P(A|B) = \frac{P(B|A) \times P(A)}{P(B)}$$

where:

* P(A|B) is the probability of event A occurring given that event B has occurred
* P(B|A) is the probability of event B occurring given that event A has occurred
* P(A) is the prior probability of event A occurring
* P(B) is the prior probability of event B occurring

#### Example
Suppose there is a medical test for a certain disease that affects 1% of the population. The test has a 95% accuracy rate, meaning that it correctly identifies 95% of people who have the disease and correctly identifies 95% of people who do not have the disease. If a person tests positive for the disease, what is the probability that they actually have the disease?

#### Using Bayes's theorem, we can calculate the probability as follows:

1. Let A be the event that a person has the disease (prior probability = 0.01)

2. Let B be the event that a person tests positive for the disease (prior probability = 0.01 x 0.95 + 0.99 x 0.05 = 0.058)

3. P(A|B) = P(B|A) x P(A) / P(B)

4. P(A|B) = 0.95 x 0.01 / 0.058

5. P(A|B) = 0.163

Therefore, the probability that a person who tests positive for the disease actually has the disease is approximately 16.3%. This calculation shows that even though the test has a high accuracy rate, there is still a significant chance of a false positive result, and additional testing or confirmation may be necessary to confirm a diagnosis.

### Q5: What is a confidence interval? How to calculate the confidence interval, explain with an example.

* A confidence interval is a range of values that is likely to contain a population parameter, such as a mean or proportion, with a certain level of confidence. It is a way of quantifying the uncertainty of an estimate based on a sample of data.

* The confidence interval is typically expressed as an interval estimate, such as "95% confidence interval for the mean is between 3.2 and 5.8". This means that if we were to repeat the sampling process many times, 95% of the time the true population mean would be within this range.

* To calculate a confidence interval, we need to know the sample size, the sample mean, and the standard error of the mean (SEM), which is a measure of the variation in the sample means. The confidence interval is based on the standard normal distribution or t-distribution, depending on the sample size and whether the population standard deviation is known.

#### Here is an example of how to calculate a confidence interval:

* Suppose we want to estimate the mean weight of a certain population of adults. We take a random sample of 100 people and measure their weights, and find that the sample mean is 72.5 kilograms and the sample standard deviation is 10.3 kilograms. We want to calculate a 99% confidence interval for the population mean weight.

* Since the sample size is greater than 30 and the population standard deviation is unknown, we use the t-distribution. The degrees of freedom is n-1, which is 99 in this case.

* First, we calculate the standard error of the mean:

    1. SEM = s / sqrt(n) = 10.3 / sqrt(100) = 1.03

    2. Next, we find the t-value for a 99% confidence level and 99 degrees of freedom using a t-table or calculator. The value is approximately 2.63.

    3. The lower and upper bounds of the confidence interval are calculated as:
        * Lower bound = sample mean - t-value x SEM = 72.5 - 2.63 x 1.03 = 69.8
        * Upper bound = sample mean + t-value x SEM = 72.5 + 2.63 x 1.03 = 75.2

* Therefore, the 99% confidence interval for the population mean weight is between 69.8 and 75.2 kilograms. This means that we are 99% confident that the true population mean weight lies within this range.

In [2]:
# Verifying above with python code
from math import sqrt
import scipy.stats as stats

# Given Below data
sample_mean = 72.5
sample_std = 10.3
n = 100

# Calculating SEM
SEM  = sample_std/sqrt(n)

# Defining the significance level 
alpha = 0.01

# Calculating the t value
t_value = stats.t.ppf(1-alpha/2, df=n-1)
print(f't_value for {n} samples with significance level of {alpha} is : {t_value:.2f}')

# Calculate the confidence interval
lower_bound = sample_mean - t_value * SEM
upper_bound = sample_mean + t_value * SEM

# Printing the results
print(f"The 99% confidence interval for population mean is ({lower_bound:.1f}, {upper_bound:.1f})")

t_value for 100 samples with significance level of 0.01 is : 2.63
The 99% confidence interval for population mean is (69.8, 75.2)


### Q6. Use Bayes' Theorem to calculate the probability of an event occurring given prior knowledge of the event's probability and new evidence. Provide a sample problem and solution.

**Suppose a company manufactures a certain type of product that has a defect rate of 2%. The company uses a quality control process that is 95% effective in catching defective products, and 99% effective in passing non-defective products. If a product passes the quality control process, what is the probability that it is actually defect-free?**


To solve this problem using Bayes' theorem, we first define the following events:

1. D: the product is defective
2. ~D: the product is not defective
3. P: the product passes the quality control process

**Solution :**
~~~
P(~D|P) = P(P|~D) * P(~D) / P(P)
P(P) = P(P|D) * P(D) + P(P|~D) * P(~D)

P(D) = 0.02
P(~D) = 0.98
P(P|D) = 0.05
P(P|~D) = 0.99

P(P) = P(P|D) * P(D) + P(P|~D) * P(~D)
     = 0.05 * 0.02 + 0.99 * 0.98
     = 0.9756

P(~D|P) = P(P|~D) * P(~D) / P(P)
        = 0.99 * 0.98 / 0.9756
        = 0.9990
~~~


**Therefore, the probability that a product is not defective given that it passes the quality control process is approximately 0.9990, or 99.90%. This result indicates that if a product passes the quality control process, it is highly likely to be defect-free. However, it is important to note that no quality control process is perfect, and some defective products may still slip through.**

In [3]:
# Solution for above with python

# Define the prior probabilities and conditional probabilities
p_defect = 0.02
p_pass_given_defect = 0.05
p_not_defect = 0.98
p_pass_given_not_defect = 0.99

# Calculate the total probability of passing the quality control process
p_pass = p_pass_given_defect * p_defect + p_pass_given_not_defect * p_not_defect

# Calculate the probability of the product being defect-free given it passed the quality control process
p_not_defect_given_pass = (p_pass_given_not_defect * p_not_defect) / p_pass

# Print the result
print(f"The probability of the product being defect-free given it passed the quality control process is {p_not_defect_given_pass:.4f}")

The probability of the product being defect-free given it passed the quality control process is 0.9990


### Q7. Calculate the 95% confidence interval for a sample of data with a mean of 50 and a standard deviation of 5. Interpret the results.

#### Assuming sample size of 30 

In [4]:
import math
import scipy.stats as stats

# Define the sample mean, standard deviation, and sample size
sample_mean = 50
sample_std_dev = 5
sample_size = 30

# Signficance Level
alpha = 0.05

# Calculate the critical value for a 95% confidence interval
t_crit = stats.t.ppf(1-alpha/2, df=sample_size-1)
print(f't statistic for significance level {alpha} , sample size of {sample_size} is : {t_crit:.2f}')

# Calculate the standard error
std_error = sample_std_dev / math.sqrt(sample_size)

# Calculate the lower and upper bounds of the confidence interval
lower_bound = sample_mean - t_crit * std_error
upper_bound = sample_mean + t_crit * std_error

# Print the results
print(f"The 95% confidence interval is ({lower_bound:.2f}, {upper_bound:.2f})")


t statistic for significance level 0.05 , sample size of 30 is : 2.05
The 95% confidence interval is (48.13, 51.87)


#### Interpretation 
From above results we can say with **95% condfidence** that the **POPULATION MEAN** Lies in between **48.13 and 51.87**.

### Q8. What is the margin of error in a confidence interval? How does sample size affect the margin of error? Provide an example of a scenario where a larger sample size would result in a smaller margin of error. 

* The margin of error in a confidence interval is the range of values around the sample statistic (such as the sample mean) within which we expect the true population parameter (such as the population mean) to lie with a certain level of confidence. 
* The margin of error is affected by the sample size, level of confidence, and standard deviation of the population.
* As the sample size increases, the margin of error decreases because larger samples provide more information about the population, and hence the sample statistics are more likely to be representative of the population. 
* The decrease in margin of error with an increase in sample size follows the square root law, which states that the margin of error decreases by the square root of the sample size.

**For example, suppose we want to estimate the average height of all students in a university using a random sample of students. We take a sample of 50 students and calculate their average height to be 170 cm with a standard deviation of 5 cm. We want to calculate a 95% confidence interval for the true population mean height.**

1. Using the formula for the margin of error for a confidence interval, we get:
~~~
Margin of error = z-value * (standard deviation / sqrt(sample size))
~~~

2. For a 95% confidence interval, the critical z-value is 1.96. So, plugging in the values, we get:
~~~
Margin of error = 1.96 * (5 / sqrt(50)) = 1.38 cm
~~~

3. This means that we are 95% confident that the true population mean height is within 1.38 cm of our sample mean of 170 cm. If we had taken a larger sample size of 200 students, the margin of error would have been:
~~~
Margin of error = 1.96 * (5 / sqrt(200)) = 0.98 cm
~~~

**So, a larger sample size would result in a smaller margin of error, making our estimate of the true population mean height more precise.**

### Q9. Calculate the z-score for a data point with a value of 75, a population mean of 70, and a population standard deviation of 5. Interpret the results.

In [5]:
# Define the values
x = 75  # Data point value
mu = 70  # Population mean
sigma = 5  # Population standard deviation

# Calculate the z-score
z = (x - mu) / sigma

print(f"The z-score is: {z}")

The z-score is: 1.0


#### Interpretation
* This means that the data point is 1 standard deviation above the population mean. 
* A positive z-score indicates that the data point is above the mean, while a negative z-score indicates that it is below the mean. 
* The magnitude of the z-score tells us how far away the data point is from the mean in terms of standard deviations.

In [6]:
from scipy.stats import norm

# Calculate the percentage of values below a z-score of 1
percent_below = norm.cdf(z) * 100

print(f"The percentage of values below a z-score of 1 is: {percent_below:.2f}%")

The percentage of values below a z-score of 1 is: 84.13%


#### Interpretation
* A z-score of 1 corresponds to approximately 84.13% of values being below this value in a standard normal distribution.

### Q10. In a study of the effectiveness of a new weight loss drug, a sample of 50 participants lost an average of 6 pounds with a standard deviation of 2.5 pounds. Conduct a hypothesis test to determine if the drug is significantly effective at a 95% confidence level using a t-test.

In [8]:
import numpy as np
from scipy.stats import t

# Null hypothesis: the drug is not significantly effective
# Alternative hypothesis: the drug is significantly effective
alpha = 0.05  # significance level
null_hypothesis = "the drug is not significantly effective"
alternative_hypothesis = "the drug is significantly effective"

mu = 0
sample_mean = 6
sample_std = 2.5
n = 50
df = n - 1  # degrees of freedom

# Calculate the t-score and p-value
t_score = (sample_mean - mu) / (sample_std / np.sqrt(n))
p_value = 2 * (1 - t.cdf(abs(t_score), df))  # or stats.t.ppf(1 - alpha/2, dof)  # two-tailed t-value

# Compare p-value with alpha and make a conclusion
print(f"t-statistic: {t_score:.4f}")
print(f"p-value: {p_value}")
if p_value < alpha:
    print(f"Reject the null hypothesis. {alternative_hypothesis}.")
else:
    print(f"Fail to reject the null hypothesis. {null_hypothesis}.")

t-statistic: 16.9706
p-value: 0.0
Reject the null hypothesis. the drug is significantly effective.


### Q11. In a survey of 500 people, 65% reported being satisfied with their current job. Calculate the 95% confidence interval for the true proportion of people who are satisfied with their job.

In [9]:
import math

p = 0.65  # sample proportion
n = 500  # sample size
ci = 0.95
alpha = 1-ci # 1-0.95 = 0.05
z_alpha = 1.96  # z-score for 95% confidence level

# Calculate the standard error
se = math.sqrt((p * (1 - p)) / n)

# Calculate the margin of error
me = z_alpha * se

# Calculate the confidence interval
lower_bound = p - me
upper_bound = p + me

# Print the results
print(f"The 95% confidence interval for the proportion of people who are satisfied with their job is ({lower_bound*100:.2f}%, {upper_bound*100:.2f}%)")

The 95% confidence interval for the proportion of people who are satisfied with their job is (60.82%, 69.18%)


### Q12. A researcher is testing the effectiveness of two different teaching methods on student performance. Sample A has a mean score of 85 with a standard deviation of 6, while sample B has a mean score of 82 with a standard deviation of 5. Conduct a hypothesis test to determine if the two teaching methods have a significant difference in student performance using a t-test with a significance level of 0.01.

In [13]:
import numpy as np
from scipy.stats import t

# Sample A
x1 = 85
s1 = 6
n1 = 30

# Sample B
x2 = 82
s2 = 5
n2 = 30

# Set up null and alternative hypotheses
# H0: mu1 = mu2 (the means are equal)
# Ha: mu1 != mu2 (the means are not equal)
alpha = 0.01
null_hypothesis = "mu1 = mu2 (the means are equal)"
alternative_hypothesis = "mu1 != mu2 (the means are not equal)"

# Calculate pooled standard deviation
sp = np.sqrt(((n1 - 1) * s1 ** 2 + (n2 - 1) * s2 ** 2) / (n1 + n2 - 2))

# Calculate t-statistic
t_stat = (x1 - x2) / (sp * np.sqrt(1 / n1 + 1 / n2))

# Calculate degrees of freedom
df = n1 + n2 - 2

# Calculate critical t-value
t_crit = t.ppf(1-alpha / 2, df)

# Calculate p-value
p_value = 2 * (1 - t.cdf(abs(t_stat), df))

# Print results
print(f"t-statistic: {t_stat:.3f}")
print(f"Degrees of freedom: {df}")
print(f"Critical t-value: {t_crit:.3f}")
print(f"p-value: {p_value:.3f}")

if abs(t_stat) > t_crit:
    print(f"Since, t_stat > t_crit i.e. {t_stat:.3f} > {t_crit:.3f} : Reject null hypothesis.")
    print(alternative_hypothesis)
else:
    print(f"Since, t_stat < t_crit i.e. {t_stat:.3f} < {t_crit:.3f} : Fail to reject null hypothesis.")
    print(null_hypothesis)

t-statistic: 2.104
Degrees of freedom: 58
Critical t-value: 2.663
p-value: 0.040
Since, t_stat < t_crit i.e. 2.104 < 2.663 : Fail to reject null hypothesis.
mu1 = mu2 (the means are equal)


### Q13. A population has a mean of 60 and a standard deviation of 8. A sample of 50 observations has a mean of 65. Calculate the 90% confidence interval for the true population mean.

In [14]:
import math
from scipy.stats import t

pop_mean = 60
pop_std = 8
n = 50
sample_mean = 65
confidence_level = 0.90
alpha = 1 - confidence_level # 1-0.90 = 0.1

# Calculate the t-score
t_score = (sample_mean - pop_mean) / (pop_std / math.sqrt(n))

# Find the critical t-value
df = n - 1
t_crit = t.ppf((1+confidence_level)/2, df)  # or t.ppf(1- alpha/2, df)
print(f't statistic for {n} samples with {confidence_level*100:.0f}% confidence is : {t_crit:.4f}')

# Calculate the margin of error
margin_of_error = t_crit * (pop_std / math.sqrt(n))

# Calculate the confidence interval
lower_bound = sample_mean - margin_of_error
upper_bound = sample_mean + margin_of_error

# Print the results
print(f"The {confidence_level*100:.0f}% confidence interval for the true population mean is ({lower_bound:.3f}, {upper_bound:.3f})")


t statistic for 50 samples with 90% confidence is : 1.6766
The 90% confidence interval for the true population mean is (63.103, 66.897)


### Q14. In a study of the effects of caffeine on reaction time, a sample of 30 participants had an average reaction time of 0.25 seconds with a standard deviation of 0.05 seconds. Conduct a hypothesis test to determine if the caffeine has a significant effect on reaction time at a 90% confidence level using a t-test.

* In this example popluation mean is not provided 
* According to researchers typical human reaction time is 200ms to 300ms
* So, assuming the population mean as 300ms or 0.3 seconds

In [15]:
import math
import scipy.stats as stats

# Null Hypothesis: Caffeine has no significant effect on reaction time
# Alternative Hypothesis: Caffeine has a significant effect on reaction time
null_hypothesis = "Caffeine has no significant effect on reaction time"
alternative_hypothesis = "Caffeine has a significant effect on reaction time"

sample_mean = 0.25  # sample mean
sample_std_dev = 0.05  # sample standard deviation
n = 30  # sample size
pop_mean = 0.3  # population mean under null hypothesis
alpha = 0.1  # 1 - significance level = 1 - 0.90 = 0.1

# Calculate the t-statistic
t_stat = (sample_mean - pop_mean) / (sample_std_dev / math.sqrt(n))

# Calculate the critical t-value
t_crit = stats.t.ppf(1-alpha/2, df=n-1)

# Calculate the confidence interval
margin_of_error = t_crit * (sample_std_dev / math.sqrt(n))
lower_ci = sample_mean - margin_of_error
upper_ci = sample_mean + margin_of_error

# Print the results
print(f"t-statistic: {t_stat:.3f}")
print(f"t-critical value: {t_crit:.3f}")
print(f"90% Confidence Interval: ({lower_ci:.3f}, {upper_ci:.3f})")

# Determine if the null hypothesis should be rejected or not
if abs(t_stat) > t_crit:
    print("Reject the null hypothesis")
    print(alternative_hypothesis)
else:
    print("Fail to reject the null hypothesis")
    print(null_hypothesis)

t-statistic: -5.477
t-critical value: 1.699
90% Confidence Interval: (0.234, 0.266)
Reject the null hypothesis
Caffeine has a significant effect on reaction time
